c/g6-read-iterator.c

```diff
@@ -0,0 +1,795 @@
+	{
+		sprintf(Line, "Invalid number of bytes for graph of order %d; got %d but expected %d\n", graphOrder, numCharsForGraphEncoding, expectedNumChars);
+		ErrorMessage(Line);
+		return NOTOK;
+	}
+
+	// Check that characters are valid ASCII characters between 62 and 126
+	for (int i = 0; i < numChars; i++)
+	{
+		if (graphBuff[i] < 63 || graphBuff[i] > 126)
+		{
+			sprintf(Line, "Invalid character at index %d: \"%c\"\n", i, graphBuff[i]);
+			ErrorMessage(Line);
+			return NOTOK;
+		}
+	}
+
+	// Check that there are no extraneous bits in representation (since we pad out to a
+	// multiple of 6 before splitting into bytes and adding 63 to each byte)
+	int expectedNumPaddingZeroes = _getExpectedNumPaddingZeroes(graphOrder, numChars);
+	char finalByte = graphBuff[numChars - 1] - 63;
+	int numPaddingZeroes = 0;
+	for (int i = 0; i < expectedNumPaddingZeroes; i++)
+	{
+		if (finalByte & (1 << i))
+			break;
+		
+		numPaddingZeroes++;
+	}
+
+	if (numPaddingZeroes != expectedNumPaddingZeroes)
+	{
+		sprintf(Line, "Expected %d padding zeroes, but got %d.\n", expectedNumPaddingZeroes, numPaddingZeroes);
+		ErrorMessage(Line);
+		exitCode = NOTOK;
+	}
+
+	return exitCode;
+}
+
+// Takes the character array graphBuff, the derived number of vertices graphOrder,
+// and the numChars corresponding to the number of characters after the first byte
+// and performs the inverse transformation of the graph encoding: we subtract 63 from
+// each byte, then only process the 6 least significant bits of the resulting byte. For
+// the final byte, we determine how many padding zeroes to expect, and exclude them
+// from being processed. We index into the adjacency matrix by row and column, which
+// are incremented such that row ranges from 0 to one less than the column index.
+int _decodeGraph(char *graphBuff, const int graphOrder, const int numChars, graphP pGraph)
+{
+	int exitCode = OK;
+
+	if (pGraph == NULL)
+	{
+		ErrorMessage("Must initialize graph datastructure before invoking _decodeGraph.\n");
+		return NOTOK;
+	}
+
+	int numPaddingZeroes = _getExpectedNumPaddingZeroes(graphOrder, numChars);
+
+	char currByte = '\0';
+	int bitValue = 0;
+	int row = 0;
+	int col = 1;
+	for (int i = 0; i < numChars; i++)
+	{
+		currByte = graphBuff[i] - 63;
+		// j corresponds to the number of places one must bitshift the byte by to read
+		// the next bit in the byte
+		for (int j = sizeof(char) * 5; j >= 0; j--)
+		{
+			// If we are on the final byte, we know that the final numPaddingZeroes bits
+			// can be ignored, so we break out of the loop
+			if ((i == numChars) && j == numPaddingZeroes - 1)
+				break;
+
+			if (row == col)
+			{
+				row = 0;
+				col++;
+			}
+
+			bitValue = ((currByte >> j) & 1u) ? 1 : 0;
+			if (bitValue == 1)
+			{
+				// Add gp_GetFirstVertex(pGraph), which is 1 if NIL == 0 (i.e. 1-based labelling) and 0 if NIL == -1 (0-based)
+				exitCode = gp_AddEdge(pGraph, row+gp_GetFirstVertex(pGraph), 0, col+gp_GetFirstVertex(pGraph), 0);
```

Comment:
Generated `K10.g6` with

```
% ./geng -h 10 45:0 > ../edge-addition-planarity-suite-fork/c/samples/K10.g6
>A ./geng -d0D9 n=10 e=45
>Z 1 graphs generated in 0.00 sec
```

And I've verified that the current mechanism to `_decodeGraph` flames out when `gp_AddEdge` returns `NONEMBEDDABLE` (defined in `graphstructures.h`) when we hit `M = 30` edges (I guess I'm kinda surprised an obstruction to planarity (i.e. a `K_{3,3}`)wasn't found after `M >= (3 * N - 5)`, since `N > 5`??)

As discussed 2024-03-06, the plan is to write a wrapper function for `gp_AddEdge`, i.e. `gp_DynamicAddEdge`, so that we don't have to duplicate this code in each of the subordinate functions of `gp_Read` and `gp_ReadFromString`.